File: codes/202602/13/dataProcessor.js
// dataProcessor.js
// 파싱된 데이터에 대한 비즈니스 로직을 수행합니다.

const { logMessage } = require('./utils');

/**
 * 주어진 데이터 배열에서 특정 컬럼의 평균을 계산합니다.
 * 컬럼의 값은 숫자로 변환 가능한 형태여야 합니다.
 * @param {Array<Object>} data - CSV 파싱 결과인 객체 배열.
 * @param {string} columnName - 평균을 계산할 숫자 컬럼의 이름.
 * @returns {number | null} 컬럼의 평균값. 계산할 수 없으면 null 반환.
 */
function calculateAverage(data, columnName) {
    logMessage(`${columnName} 컬럼의 평균 계산을 시작합니다.`);
    if (!data || data.length === 0) {
        logMessage('데이터가 비어 있어 평균을 계산할 수 없습니다.');
        return null;
    }

    let sum = 0;
    let count = 0;

    for (const row of data) {
        const value = parseFloat(row[columnName]);
        if (!isNaN(value)) {
            sum += value;
            count++;
        } else {
            logMessage(`경고: '${row[columnName]}' (컬럼: ${columnName}) 값은 숫자가 아니므로 평균 계산에서 제외됩니다.`);
        }
    }

    if (count === 0) {
        logMessage(`'${columnName}' 컬럼에 유효한 숫자 값이 없어 평균을 계산할 수 없습니다.`);
        return null;
    }

    const average = sum / count;
    logMessage(`'${columnName}' 컬럼의 평균: ${average}`);
    return average;
}

module.exports = {
    calculateAverage
};

File: codes/202602/26/DataProcessor.cs
// DataProcessor.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CsvProcessor
{
    /// <summary>
    /// 데이터를 처리하는 다양한 유틸리티 메서드를 제공하는 클래스입니다.
    /// </summary>
    public static class DataProcessor
    {
        /// <summary>
        /// 지정된 컬럼의 숫자 값들에 대한 평균을 계산합니다.
        /// </summary>
        /// <param name="records">처리할 데이터 레코드 리스트입니다.</param>
        /// <param name="columnName">평균을 계산할 컬럼의 이름입니다.</param>
        /// <returns>지정된 컬럼의 평균 값입니다.</returns>
        public static double CalculateAverage(List<Dictionary<string, string>> records, string columnName)
        {
            if (records == null || !records.Any())
            {
                return 0.0;
            }

            double sum = 0.0;
            int count = 0;

            foreach (var record in records)
            {
                if (record.TryGetValue(columnName, out string valueString) && double.TryParse(valueString, out double value))
                {
                    sum += value;
                    count++;
                }
            }

            return count > 0 ? sum / count : 0.0;
        }

        /// <summary>
        /// 지정된 컬럼의 값이 특정 기준 값과 일치하는 레코드들을 필터링합니다.
        /// </summary>
        /// <param name="records">필터링할 데이터 레코드 리스트입니다.</param>
        /// <param name="columnName">필터링 기준이 될 컬럼의 이름입니다.</param>
        /// <param name="filterValue">컬럼 값이 일치해야 하는 기준 값입니다.</param>
        /// <returns>필터링된 레코드 리스트입니다.</returns>
        public static List<Dictionary<string, string>> FilterData(List<Dictionary<string, string>> records, string columnName, string filterValue)
        {
            if (records == null || !records.Any())
            {
                return new List<Dictionary<string, string>>();
            }

            return records.Where(record =>
                record.TryGetValue(columnName, out string value) && value.Equals(filterValue, StringComparison.OrdinalIgnoreCase))
                .ToList();
        }
    }
}